src/suggestions.ts

```diff
@@ -0,0 +1,75 @@
+import fs = require("fs");
+import os = require("os");
+import path = require("path");
+import { WalkContext } from "tslint";
+
+const suggestionsDir = path.join(os.homedir(), ".dts", "suggestions");
+
+export interface Suggestion {
+    fileName: string;
+    ruleName: string;
+    message: string;
+    start?: number;
+    width?: number;
+}
+
+// Packages for which suggestions were already added in this run of dtslint.
+const existingPackages = new Set();
+
+/**
+ *  A rule should call this function to provide a suggestion instead of a lint failure.
+ */
+export function addSuggestion<T>(ctx: WalkContext<T>, message: string, start?: number, width?: number) {
+    const suggestion: Suggestion = {
+        fileName: ctx.sourceFile.fileName,
+        ruleName: ctx.ruleName,
+        message,
+        start,
+        width,
+    };
+
+    const packageName = dtPackageName(ctx.sourceFile.fileName);
+    if (!packageName) {
+        return;
+    }
+    let flag = "a";
+    if (!existingPackages.has(packageName)) {
+        flag = "w";
+        existingPackages.add(packageName);
+    }
+    try {
+        if (!fs.existsSync(suggestionsDir)) {
+            fs.mkdirSync(suggestionsDir, { recursive: true });
+        }
+        fs.writeFileSync(
+            path.join(suggestionsDir, packageName + ".txt"),
```

Comment:
I'd recommend writing a single json. Either you can write the suggestions all at once or look for an existing file and read it.

1. let x = JSON.parse(readFileSync(packageName + ".json"))
2. x[packageName] = formatSuggestion(suggestion)

Probably writing all the suggestions into a single JSON at the end is clearer in addition to being more efficient.

Reply:
As discussed offline, since `addSuggestion` is called per file, we don't have all the suggestions for a package at once, so it's complicated to write all the suggestions to a file a single time.